jax/_src/nn/functions.py

```diff
@@ -765,3 +768,203 @@ def hard_silu(x: ArrayLike) -> Array:
+                `probs`.
+    _qk_dot_general: custom dot function for qk dot (mainly for quantization).
+    _pv_dot_general: custom dot function for pv dot (mainly for quantization).
+
+  Returns:
+    An array with the same shape of :code:`query`. If flash attention is not
+    used, the `probs` array will also be returned.
+  """
+  def _assert_has_shape(t: ArrayLike, shape: Sequence[int]) -> None:
+    assert t.ndim == len(shape)
+    value_str1 = f't.shape={t.shape}'
+    value_str2 = f'shape={shape}'
+    for i in range(t.ndim):
+      if shape[i] != -1:
+        assert (
+            t.shape[i] == shape[i]
+        ), f'Shapes are not equal {value_str1} vs {value_str2}'
+    
+  B, S, N, H = key.shape
+  _assert_has_shape(value, [B, S, N, H])
+  _assert_has_shape(query, [B, -1, N, H])
+  T = query.shape[1]
+  scale_val = (1.0 / np.sqrt(H) if scale is None else scale)
+
+  if not (mask is None or mask_fn is None):
+      raise ValueError("Cannot provide both mask and mask_fn")
+
+  # Try if the flash attention can be used. If failed, fall back to the default 
+  # implementation.
+  try:
+    maybe_fuse = softmax_fn is None
+    maybe_fuse = maybe_fuse and (dropout_fn is None or
+                                 type(dropout_fn) is SdpaPhiloxDropout)
+    maybe_fuse = maybe_fuse and (mask_fn is None or
+                                 type(mask_fn) is SdpaCausalMask)
+    if maybe_fuse:
+      rate, seed = 0.0, 0
+      if type(dropout_fn) is SdpaPhiloxDropout:
+        rate, seed = (dropout_fn.rate, dropout_fn.seed)
+
+      mask_type = MaskType.NO_MASK
+      if type(mask_fn) is SdpaCausalMask:
+        mask_type = MaskType.CAUSAL
+
+      # The `mask` to this API is an additive mask. However, the `mask` taken by
+      # the dot_product_attention is assumed to be a multiplicative
+      # mask. So, we pass it via `bias` to preserve its "additive" behavior.
+      if bias is not None:
+        # Cudnn flash attention applies post-scale bias. So we scale the bias.
+        # Note, mathematically, (logits+bias)*scale is equivalent to
+        # logits*scale+bias*scale, but the latter may cause numerical
+        # differences especially when bias is in narrower dtypes.
+        scale = jnp.array(scale_val, dtype=bias.dtype)
+        bias = jnp.multiply(bias, scale)
+        if mask is not None:
+          bias = bias + mask
+      elif mask is not None:
+        bias = mask
+
+      encoded = dot_product_attention(
+          query, key, value, bias, None, scale=scale_val, mask_type=mask_type,
+          seed=seed, dropout_rate=rate
+      )
+      return encoded, None
+    else:
+      warnings.warn("The flash attention cannot be used because unsupported"
+                    " softmax_fn, mask_fn, or dropout_fn.")
+  except Exception as e:
+    warnings.warn(f"The flash attention cannot be used because: {e}")
+
+  # Compute the attention logits
+  logits = jnp.einsum('BTNH,BSNH->BNTS', query, key,
```

Comment:
In the transformer engine docs (https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/jax.html#transformer_engine.jax.flax.DotProductAttention) for the JAX cuDNN fused attention API, it states "float32_logits (bool, default = False) – Whether to compute attention logits in float32 for the unfused attention backend. For fused attention backend, the accumulation is always float32 without the perf overhead." It would be good to match numerics as closely as possible. What exactly is cuDNN doing here? Is it accumulating in FP32 (say for bf16 inputs) and then keeping it in FP32 to compute softmax? Or does it cast the accumulation to BF16 like this does?

Reply:
Yes, we accumulate in fp32 and do the softmax directly on it. We don't do a cast_to BF16 and then cast_back_to FP32 for softmax in flash attention.

Reply:
OK so you are using quite different numerics for BF16 in your XLA and cuDNN implementations. The XLA-equivalent-here is to use `preferred_element_type=jnp.float32` in `jnp.einsum`. This has the negative consequence of using much more memory. But having inconsistent numerics might mean that if you train with cuDNN, inference will no longer work due to the downcasted numerics. Is there an option for controlling this in cuDNN, so we can do the XLA-default downcast behaviour by default and users can switch using `preferred_element_type=jnp.float32`?

Reply:
Cudnn uses a different algorithm for the attention (i.e. the flash attention algorithm). So, from it alone, the numerics would be very different from XLA. As for the dot compute/accumulation dtype, I don't think the current cudnn allow a config for that. 

@Cjkkkk Can you comment on this? Basically, I think google's concern is how to best match the numerics from the default other implementations. Or do you know if we have tried things like train model with one attention but inference with another?

Reply:
FlashAttention is numerically identical to XLA attention, modulo doing ops in different orders, which is common in ML. Here we are discussing downcasting the logits to BF16 versus not downcasting to BF16. We should match numerics. There are two options:
- Make the cuDNN numerics the default, and match this in XLA (so use `preferred_element_type=jnp.float32` in the einsum). Comes at the cost of larger default XLA memory usage and slower speed (more memory to move). Allow users to opt-out of this with an option (can add this later).
- Make downcasting the default. Need option then to use cuDNN attention. 

I think I prefer the first option. Either way, should document this.

Reply:
Ah, I see your point now. I think we did something similar in Transformer Engine and some arg like https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/softmax.py#L274. So, we can add a similar one in the API and set it to true by default. Then it's like:
```
xla:
# softmax_in_fp32=True (default)
preferred_element_type = fp32 if softmax_in_fp32 else None
jnp.einsum(..., preferred_element_type=preferred_element_type)

cudnn:
if softmax_in_fp32:
  raise some error.
```
What do you think?

Reply:
the softmax in flash attention (online softmax) is fundamentally different algorithms, mathematically should be equivalent to JAX softmax but due to floating point computation, it is not numerically the same.

Reply:
cuDNN attention is doing accumulation in fp32 and no downcasting to bf16 before softmax so the first option would align both attentions.

Reply:
Added the flag `softmax_in_fp32` and defaulted it to `True`.